#! /bin/sh /usr/share/dpatch/dpatch-run
## 81_sonames.dpatch by Mike Hommey <glandium@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Add soname support

@DPATCH@

--- nss/mozilla/security/coreconf/BSD_OS.mk
+++ nss/mozilla/security/coreconf/BSD_OS.mk
@@ -73,7 +73,7 @@
 DSO_CFLAGS		= -fPIC -DPIC
 DSO_LDOPTS		= -shared
 DSO_LDFLAGS		=
-DSO_LDOPTS		+= -Wl,-soname,lib$(LIBRARY_NAME)$(LIBRARY_VERSION).$(DLL_SUFFIX)
+DSO_LDOPTS		+= -Wl,-soname,$(SONAME)
 
 ifdef LIBRUNPATH
 DSO_LDOPTS		+= -Wl,-R$(LIBRUNPATH)
--- nss/mozilla/security/coreconf/BeOS.mk
+++ nss/mozilla/security/coreconf/BeOS.mk
@@ -59,7 +59,7 @@
 	CPU_ARCH	= x86
 endif
 
-MKSHLIB		= $(CC) -nostart -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB		= $(CC) -nostart -Wl,-soname -Wl,$(SONAME)
 ifdef BUILD_OPT
 	OPTIMIZER	= -O2
 endif
--- nss/mozilla/security/coreconf/FreeBSD.mk
+++ nss/mozilla/security/coreconf/FreeBSD.mk
@@ -51,7 +51,7 @@
 OS_CFLAGS		= $(DSO_CFLAGS) -ansi -Wall -DFREEBSD -DHAVE_STRERROR -DHAVE_BSD_FLOCK
 
 DSO_CFLAGS		= -fPIC
-DSO_LDOPTS		= -shared -Wl,-soname -Wl,$(notdir $@)
+DSO_LDOPTS		= -shared -Wl,-soname -Wl,$(SONAME)
 
 #
 # The default implementation strategy for FreeBSD is pthreads.
--- nss/mozilla/security/coreconf/HP-UX.mk
+++ nss/mozilla/security/coreconf/HP-UX.mk
@@ -83,7 +83,7 @@
 PROCESS_MAP_FILE = grep -v ';+' $< | grep -v ';-' | \
          sed -e 's; DATA ;;' -e 's,;;,,' -e 's,;.*,,' -e 's,^,+e ,' > $@
 
-DSO_LDOPTS		= -b +h $(notdir $@)
+DSO_LDOPTS		= -b +h $(SONAME)
 ifeq ($(OS_TEST),ia64)
 	DSO_LDOPTS	+= +b '$$ORIGIN'
 endif
--- nss/mozilla/security/coreconf/IRIX.mk
+++ nss/mozilla/security/coreconf/IRIX.mk
@@ -102,7 +102,7 @@
 	SHLIB_LD_OPTS	+= -n32 -mips3
 endif
 
-MKSHLIB     += $(LD) $(SHLIB_LD_OPTS) -shared -soname $(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB     += $(LD) $(SHLIB_LD_OPTS) -shared -soname $(SONAME)
 ifdef MAPFILE
 # Add LD options to restrict exported symbols to those in the map file
 endif
--- nss/mozilla/security/coreconf/Linux.mk
+++ nss/mozilla/security/coreconf/Linux.mk
@@ -139,7 +139,7 @@
 
 ifeq ($(KERNEL)-$(OS_RELEASE),linux-2.0)
 	OS_REL_CFLAGS	+= -DLINUX2_0
-	MKSHLIB		= $(CC) -shared -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+	MKSHLIB		= $(CC) -shared -Wl,-soname -Wl,$(SONAME)
 	ifdef MAPFILE
 		MKSHLIB += -Wl,--version-script,$(MAPFILE)
 	endif
--- nss/mozilla/security/coreconf/Linux2.1.mk
+++ nss/mozilla/security/coreconf/Linux2.1.mk
@@ -38,7 +38,7 @@
 include $(CORE_DEPTH)/coreconf/Linux.mk
 ifeq ($(OS_RELEASE),2.1)
         OS_REL_CFLAGS   += -DLINUX2_1
-        MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+        MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(SONAME)
 	ifdef MAPFILE
 		MKSHLIB += -Wl,--version-script,$(MAPFILE)
 	endif
--- nss/mozilla/security/coreconf/Linux2.2.mk
+++ nss/mozilla/security/coreconf/Linux2.2.mk
@@ -38,7 +38,7 @@
 include $(CORE_DEPTH)/coreconf/Linux.mk
 
 OS_REL_CFLAGS   += -DLINUX2_1
-MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(SONAME)
 
 ifdef MAPFILE
 	MKSHLIB += -Wl,--version-script,$(MAPFILE)
--- nss/mozilla/security/coreconf/Linux2.4.mk
+++ nss/mozilla/security/coreconf/Linux2.4.mk
@@ -38,7 +38,7 @@
 include $(CORE_DEPTH)/coreconf/Linux.mk
 
 OS_REL_CFLAGS   += -DLINUX2_1
-MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(SONAME)
 
 ifdef MAPFILE
 	MKSHLIB += -Wl,--version-script,$(MAPFILE)
--- nss/mozilla/security/coreconf/Linux2.5.mk
+++ nss/mozilla/security/coreconf/Linux2.5.mk
@@ -38,7 +38,7 @@
 include $(CORE_DEPTH)/coreconf/Linux.mk
 
 OS_REL_CFLAGS   += -DLINUX2_1
-MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(SONAME)
 
 ifdef MAPFILE
 	MKSHLIB += -Wl,--version-script,$(MAPFILE)
--- nss/mozilla/security/coreconf/Linux2.6.mk
+++ nss/mozilla/security/coreconf/Linux2.6.mk
@@ -42,7 +42,7 @@
 OS_REL_CFLAGS   += -DLINUX2_1
 endif

-MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB         = $(CC) $(DSO_LDOPTS) -Wl,-soname -Wl,$(SONAME)
 
 ifdef MAPFILE
 	MKSHLIB += -Wl,--version-script,$(MAPFILE)
--- nss/mozilla/security/coreconf/NetBSD.mk
+++ nss/mozilla/security/coreconf/NetBSD.mk
@@ -68,7 +68,7 @@
 DSO_LDOPTS		= -shared
 DSO_LDFLAGS		=
 ifeq ($(OBJECT_FMT),ELF)
-DSO_LDOPTS		+= -Wl,-soname,lib$(LIBRARY_NAME)$(LIBRARY_VERSION).$(DLL_SUFFIX)
+DSO_LDOPTS		+= -Wl,-soname,$(SONAME)
 endif
 
 ifdef LIBRUNPATH
--- nss/mozilla/security/coreconf/OSF1.mk
+++ nss/mozilla/security/coreconf/OSF1.mk
@@ -64,7 +64,7 @@
 endif
 
 # The command to build a shared library on OSF1.
-MKSHLIB    += ld -shared -expect_unresolved "*" -soname $(notdir $@)
+MKSHLIB    += ld -shared -expect_unresolved "*" -soname $(SONAME)
 ifdef MAPFILE
 MKSHLIB += -hidden -input $(MAPFILE)
 endif
--- nss/mozilla/security/coreconf/OpenBSD.mk
+++ nss/mozilla/security/coreconf/OpenBSD.mk
@@ -64,7 +64,7 @@
 ARCH			= openbsd
 
 DSO_CFLAGS		= -fPIC -DPIC
-DSO_LDOPTS		= -shared -fPIC -Wl,-soname,lib$(LIBRARY_NAME)$(LIBRARY_VERSION).$(DLL_SUFFIX)
+DSO_LDOPTS		= -shared -fPIC -Wl,-soname,$(SONAME)
 DSO_LDFLAGS		=
 
 MKSHLIB			= $(CC) $(DSO_LDOPTS)
--- nss/mozilla/security/coreconf/QNX.mk
+++ nss/mozilla/security/coreconf/QNX.mk
@@ -54,7 +54,7 @@
 	CPU_ARCH	= x86
 endif
 
-MKSHLIB		= $(CC) -shared -Wl,-soname -Wl,$(@:$(OBJDIR)/%.so=%.so)
+MKSHLIB		= $(CC) -shared -Wl,-soname -Wl,$(SONAME)
 ifdef BUILD_OPT
 	OPTIMIZER	= -O2
 endif
--- nss/mozilla/security/coreconf/ReliantUNIX.mk
+++ nss/mozilla/security/coreconf/ReliantUNIX.mk
@@ -51,7 +51,7 @@
 		OPTIMIZER += -O6
 	endif
 	MKSHLIB     = $(LD)
-	MKSHLIB    += -G -h $(@:$(OBJDIR)/%.so=%.so)
+	MKSHLIB    += -G -h $(SONAME)
 	DSO_LDOPTS += -G -Xlinker -Blargedynsym
 else
 	## native compiler (CDS++ 1.0)
--- nss/mozilla/security/coreconf/SunOS5.mk
+++ nss/mozilla/security/coreconf/SunOS5.mk
@@ -170,7 +170,7 @@
 ifeq ($(USE_64), 1)
 	DSO_LDOPTS += -m64
 endif
-	DSO_LDOPTS += -shared -h $(notdir $@)
+	DSO_LDOPTS += -shared -h $(SONAME)
 else
 ifeq ($(USE_64), 1)
 	ifeq ($(OS_TEST),i86pc)
@@ -179,7 +179,7 @@
 	    DSO_LDOPTS +=-xarch=v9
 	endif
 endif
-	DSO_LDOPTS += -G -h $(notdir $@)
+	DSO_LDOPTS += -G -h $(SONAME)
 endif
 DSO_LDOPTS += -z combreloc -z defs -z ignore

--- nss/mozilla/security/coreconf/rules.mk
+++ nss/mozilla/security/coreconf/rules.mk
@@ -119,6 +119,7 @@
 	$(INSTALL) -m 644 $(SHARED_LIBRARY:$(DLL_SUFFIX)=pdb) $(SOURCE_LIB_DIR)
 endif
 endif
+	$(call MKSHLINKS,$(SOURCE_LIB_DIR))
 endif
 ifdef IMPORT_LIBRARY
 	$(INSTALL) -m 775 $(IMPORT_LIBRARY) $(SOURCE_LIB_DIR)
@@ -268,6 +269,7 @@
 release_md::
 ifneq ($(MD_LIB_RELEASE_FILES),)
 	$(INSTALL) -m 444 $(MD_LIB_RELEASE_FILES) $(SOURCE_RELEASE_PREFIX)/$(SOURCE_RELEASE_LIB_DIR)
+	$(call MKSHLINKS,$(SOURCE_RELEASE_PREFIX)/$(SOURCE_RELEASE_LIB_DIR))
 endif
 ifneq ($(MD_BIN_RELEASE_FILES),)
 	$(INSTALL) -m 555 $(MD_BIN_RELEASE_FILES) $(SOURCE_RELEASE_PREFIX)/$(SOURCE_RELEASE_BIN_DIR)
@@ -328,6 +330,8 @@
 endif
 endif
 
+$(SHARED_LIBRARY_LINKS): %: $(SHARED_LIBRARY)
+
 $(SHARED_LIBRARY): $(OBJS) $(RES) $(MAPFILE) $(SUB_SHLOBJS)
 	@$(MAKE_OBJDIR)
 	rm -f $@
--- nss/mozilla/security/coreconf/ruleset.mk
+++ nss/mozilla/security/coreconf/ruleset.mk
@@ -123,6 +123,25 @@
     ifndef SHARED_LIBRARY
 	SHARED_LIBRARY = $(OBJDIR)/$(DLL_PREFIX)$(LIBRARY_NAME)$(LIBRARY_VERSION)$(JDK_DEBUG_SUFFIX).$(DLL_SUFFIX)
     endif
+    SONAME = $(notdir $(SHARED_LIBRARY))
+    ifdef SO_VERSION
+    ifneq (,$(findstring $(SONAME),$(MKSHLIB)))
+        SO_VERSION_MAJOR	:= $(shell echo $(SO_VERSION) | sed 's/^\([^.]*\)\(\.[^.]*\)\?\(\.[^.]*\)\?/\1/')
+        SO_VERSION_MINOR	:= $(shell echo $(SO_VERSION) | sed 's/^\([^.]*\)\(\.[^.]*\)\?\(\.[^.]*\)\?/\2/')
+        SO_VERSION_MICRO	:= $(shell echo $(SO_VERSION) | sed 's/^\([^.]*\)\(\.[^.]*\)\?\(\.[^.]*\)\?/\3/')
+
+        SHARED_LIBRARY_LINKS	:= $(SONAME)
+        ifdef SO_VERSION_MINOR
+            SHARED_LIBRARY_LINKS	+= $(SONAME).$(SO_VERSION_MAJOR)
+        endif
+
+        SONAME			:= $(SONAME).$(SO_VERSION_MAJOR)
+        SHARED_LIBRARY		:= $(SHARED_LIBRARY).$(SO_VERSION)
+
+        MKSHLINKS		= (cd $(1) && for link in $(SHARED_LIBRARY_LINKS); do rm -f $$link; ln -s $(notdir $(SHARED_LIBRARY)) $$link; done)
+    endif
+    endif
+
     ifndef MAPFILE_SOURCE
 	MAPFILE_SOURCE = $(LIBRARY_NAME).def
     endif
@@ -177,7 +196,7 @@
 endif
 
 ALL_TRASH =	$(TARGETS) $(OBJS) $(OBJDIR) LOGS TAGS $(GARBAGE) \
-		so_locations $(BUILT_SRCS) $(NOSUCHFILE)
+		so_locations $(BUILT_SRCS) $(NOSUCHFILE) $(SHARED_LIBRARY_LINKS)
 
 ifdef NS_USE_JDK
     ALL_TRASH += $(JDK_HEADER_CFILES) $(JDK_STUB_CFILES) \
--- nss/mozilla/security/nss/lib/nss/manifest.mn
+++ nss/mozilla/security/nss/lib/nss/manifest.mn
@@ -57,6 +57,7 @@
 
 LIBRARY_NAME = nss
 LIBRARY_VERSION = 3
+SO_VERSION = 1d
 
 # This part of the code, including all sub-dirs, can be optimized for size
 export ALLOW_OPT_CODE_SIZE = 1
--- nss/mozilla/security/nss/lib/smime/manifest.mn
+++ nss/mozilla/security/nss/lib/smime/manifest.mn
@@ -79,6 +79,7 @@
 
 LIBRARY_NAME = smime
 LIBRARY_VERSION = 3
+SO_VERSION = 1d
 
 # This part of the code, including all sub-dirs, can be optimized for size
 export ALLOW_OPT_CODE_SIZE = 1
--- nss/mozilla/security/nss/lib/ssl/manifest.mn
+++ nss/mozilla/security/nss/lib/ssl/manifest.mn
@@ -86,6 +86,7 @@
 
 LIBRARY_NAME = ssl
 LIBRARY_VERSION = 3
+SO_VERSION = 1d
 
 # This part of the code, including all sub-dirs, can be optimized for size
 export ALLOW_OPT_CODE_SIZE = 1
